operators/pkg/instctrl/virtualmachines.go

```diff
@@ -83,7 +84,15 @@ func (r *InstanceReconciler) enforceVirtualMachine(ctx context.Context) error {
 	}
 	log.V(utils.FromResult(res)).Info("virtualmachine enforced", "virtualmachine", klog.KObj(&vm), "result", res)
 
-	phase := r.RetrievePhaseFromVM(&vm)
+	vmiLookupKey := types.NamespacedName{Name: vm.ObjectMeta.Name}
+	var vmi virtv1.VirtualMachineInstance
+	var phase clv1alpha2.EnvironmentPhase
+	if err = r.Get(ctx, vmiLookupKey, &vmi); err != nil {
+		phase = r.RetrievePhaseFromVM(&vm, nil)
+	} else {
+		phase = r.RetrievePhaseFromVM(&vm, &vmi)
+	}
```

Comment:
```suggestion
	// It is necessary to retrieve the VMI object associated with the VM (if any), to correctly detect the ResourceQuotaExceeded phase.
	// VM and VMI are characterized by the same resource name.
	vmi := virtv1.VirtualMachineInstance{ObjectMeta: forge.ObjectMeta(instance)}
	if err = r.Get(ctx, client.ObjectKeyFromObject(&vmi), &vmi); client.IgnoreNotFound(err) != nil {
		log.Error(err, "failed to retrieve virtualmachineinstance", "virtualmachineinstance", klog.KObj(&vm))
		return err
	}

	phase := r.RetrievePhaseFromVM(&vm, &vmi)
```

Your code had two main problems:
* It did not consider the case of possible errors, which should not be ignored
* The VMI lookup key was not correct (the namespace was missing, and the logic could not work).

BTW, the nil check in `RetrievePhaseFromVM` can be deleted, as no longer useful.